File: backend/domain/image_environment.go
package domain

import (
	"fmt"
	"unicode"
	"unicode/utf8"
)

// ImageEnvironment is administrator-declared documentation, not a platform
// attestation. ValidationNotes must never be interpreted as verified status.
type ImageEnvironment struct {
	Python          string `json:"python,omitempty"`
	CUDA            string `json:"cuda,omitempty"`
	PyTorch         string `json:"pytorch,omitempty"`
	MLflow          string `json:"mlflow,omitempty"`
	Dependencies    string `json:"dependencies,omitempty"`
	UseCases        string `json:"useCases,omitempty"`
	ValidationNotes string `json:"validationNotes,omitempty"`
}

func (e ImageEnvironment) Validate() error {
	for _, field := range []struct {
		name, value string
		limit       int
		multiline   bool
	}{
		{"python", e.Python, 128, false}, {"cuda", e.CUDA, 128, false},
		{"pytorch", e.PyTorch, 128, false}, {"mlflow", e.MLflow, 128, false},
		{"dependencies", e.Dependencies, 12000, true}, {"useCases", e.UseCases, 2000, true},
		{"validationNotes", e.ValidationNotes, 4000, true},
	} {
		if !utf8.ValidString(field.value) || utf8.RuneCountInString(field.value) > field.limit {
			return fmt.Errorf("image environment %s must be valid text of at most %d characters", field.name, field.limit)
		}
		for _, r := range field.value {
			if unicode.IsControl(r) && !(field.multiline && (r == '\n' || r == '\r' || r == '\t')) {
				return fmt.Errorf("image environment %s contains unsupported control characters", field.name)
			}
		}
	}
	return nil
}

File: frontend/src/imageEnvironment.js
export const imageEnvironmentFields = Object.freeze([
  { key: 'python', label: 'Python', max: 128 },
  { key: 'cuda', label: 'CUDA', max: 128 },
  { key: 'pytorch', label: 'PyTorch', max: 128 },
  { key: 'mlflow', label: 'MLflow', max: 128 },
  { key: 'dependencies', label: '主要依赖（含版本）', max: 12000 },
  { key: 'useCases', label: '适用模型 / 场景', max: 2000 },
  { key: 'validationNotes', label: '验证记录（命令、日期与结果）', max: 4000 },
])

export function emptyImageEnvironment() {
  return Object.fromEntries(imageEnvironmentFields.map(({ key }) => [key, '']))
}

export function imageEnvironmentRows(image = {}) {
  return [
    { key: 'ray', label: 'Ray', value: image.rayVersion || '未提供' },
    ...imageEnvironmentFields.map(({ key, label }) => ({
      key, label, value: typeof image.environment?.[key] === 'string' && image.environment[key].trim()
        ? image.environment[key] : '未提供',
    })),
  ]
}
